lib/livepush.ts

```diff
@@ -0,0 +1,112 @@
+import * as Bluebird from 'bluebird';
+import * as Dockerode from 'dockerode';
+import { EventEmitter } from 'events';
+import * as _ from 'lodash';
+import StrictEventEmitter from 'strict-event-emitter-types';
+
+import Container, { CommandOutput, StageContainers } from './container';
+import Dockerfile from './dockerfile';
+import { InvalidArgumentError } from './errors';
+
+export interface LivepushEvents {
+	commandExecute: { stageIdx: number; command: string };
+	commandOutput: { stageIdx: number; output: CommandOutput };
+	commandReturn: { stageIdx: number; returnCode: number };
+	containerRestart: { containerId: string };
+}
+
+type ContainerEventEmitter = StrictEventEmitter<EventEmitter, LivepushEvents>;
+
+export class Livepush extends (EventEmitter as {
+	// We need to avoid the tslint errors here, as typescript
+	// will not accept the changes proposed
+	// tslint:disable-next-line
+	new (): ContainerEventEmitter;
+}) {
+	private constructor(
+		public docker: Dockerode,
+		public dockerfile: Dockerfile,
+		public containers: StageContainers,
+	) {
+		super();
+		this.assignEventHandlers();
+	}
+
+	public static async init(
+		dockerfileContent: string | Buffer,
+		context: string,
+		containerId: string,
+		stageImages: string[],
+		docker: Dockerode,
+	): Promise<Livepush> {
+		const dockerfile = new Dockerfile(dockerfileContent);
+
+		if (dockerfile.stages.length - 1 !== stageImages.length) {
+			const dStages = dockerfile.stages.length;
+			const argStages = stageImages.length;
+			throw new InvalidArgumentError(
+				`Dockerfile with ${dStages} stages provided,` +
+					` but ${argStages} image IDs passed to livepush constructor (there should be ${dStages -
+						1})`,
```

Comment:
strange indentation, it this `prettier`?

Reply:
It looks horrendous; therefore, it is... :/

Reply:
Yup, all prettier.